packages/api/src/module/write/shared/application/events-subscription/ordered-event-queue.spec.ts

```diff
@@ -0,0 +1,202 @@
+}
+
+async function producer(sut: OrderedEventQueue, event: ApplicationEvent, waitingTime: number) {
+  await wait(waitingTime);
+  sut.push(event);
+}
+
+async function consumer(sut: OrderedEventQueue, from: number, waitingTime: number) {
+  const result = [];
+  let globalOrder = from;
+  let event = await sut.pop();
+
+  while (!OrderedEventQueue.isStopToken(event)) {
+    if (event.globalOrder === globalOrder) {
+      result.push(event);
+      globalOrder += 1;
+    } else {
+      sut.push(event);
+      await wait(waitingTime);
+    }
+
+    event = await sut.pop();
+  }
+
+  return result;
+}
+
+function stoppedQueueWithSomeEvents() {
+  const sut = new OrderedEventQueue();
+
+  sut.stop();
+
+  const pushedEvents = sampleEventsInRange(1, 5);
+
+  pushAll(sut, pushedEvents);
+
+  return { sut, pushedEvents };
+}
+
+describe('OrderedEventQueue', () => {
+  it('Given queue when push() then queue should be always sorted in ascending order', async () => {
+    // Given
+    const sut = new OrderedEventQueue();
+
+    const eventsBatch0 = randomEventsInRange(0, 5);
+
+    pushAll(sut, eventsBatch0);
+
+    // When
+    const poppedEvents0 = await popAll(sut);
+
+    // Then
+    expect(poppedEvents0).toStrictEqual(sorted(eventsBatch0));
+
+    // When
+    const eventsBatch1 = randomEventsInRange(5, 10);
+
+    pushAll(sut, eventsBatch1);
+    pushAll(sut, eventsBatch0);
+
+    const poppedEvents1 = await popAll(sut);
+
+    // Then
+    expect(poppedEvents1).toStrictEqual(sorted([...eventsBatch0, ...eventsBatch1]));
+  });
+
+  it('Queue should be eventualy sorted in ascending order', () => {
+    // Given
+    const sut = new OrderedEventQueue();
+    const startingIndex = 1;
+    const timeout = 1000;
+    const eventsBatch = randomEventsInRange(startingIndex, 11);
+
+    // When
+    const producers = [
+      ...eventsBatch.map((event, i) => producer(sut, event, i * 100)),
+      wait(timeout).then(() => sut.stop()),
+    ];
+
+    // Then
+    return Promise.all([
+      ...producers,
+      consumer(sut, startingIndex, 50).then((poppedEvents) => expect(poppedEvents).toStrictEqual(sorted(eventsBatch))),
+    ]);
+  });
+
+  it('Given empty queue when push() then pop() should return new event', () => {
+    // Given
+    const sut = new OrderedEventQueue();
+    const waitingTime = 1000;
+    const sampleEvent = sampleApplicationEvent({});
+
+    return Promise.all([
+      // When
+      wait(waitingTime).then(() => sut.push(sampleEvent)),
+
+      // Then
+      expect(sut.pop())
+        .resolves.toBe(sampleEvent)
+        .then(() => expect(sut.length).toBe(0)),
+    ]);
+  });
+
+  it('Given empty queue when stop() then pop() should return STOP_TOKEN', () => {
+    // Given
+    const sut = new OrderedEventQueue();
+    const timeout = 1000;
+
+    return Promise.all([
+      // When
+      wait(timeout).then(() => sut.stop()),
+      // Then
+      expect(sut.pop()).resolves.toBe(OrderedEventQueue.STOP_TOKEN),
+    ]);
```

Comment:
What do you think about? Is it the same?

```suggestion
  it('Given empty queue when stop() then pop() should return STOP_TOKEN', async () => {
    // Given
    const sut = new OrderedEventQueue();
    const timeout = 1000;

     // When
     await wait(timeout).then(() => sut.stop()),
     // Then
     await expect(sut.pop()).resolves.toBe(OrderedEventQueue.STOP_TOKEN),
```

Reply:
No, because after this `await wait(timeout).then(() => sut.stop()),` pop will be resolved. I wanna test the fact that `pop()` blocks on promise and `stop()` after some period of time unblocks that `promise` with `STOP_TOKEN`.

Reply:
Ah, now I understand :) Co maybe 'concurrently' in test description, or something like that will reveal the test motivations?